File: books/cs/introduction-to-algorithms/chpater3/Ruby/counting_sort.rb

dataset_size = 100;
max_value = 100;
aaa = Array.new(dataset_size);
bbb = Array.new(dataset_size);
ccc = [];
k = 0;
idx = 0;
total = 0;

0.upto(dataset_size-1) do |idx|
	aaa[idx] = rand(max_value)
end;
puts aaa;

k = aaa.max;
ccc = Array.new(k+1);
0.upto(k){|idx| ccc[idx]=0;}
puts "#{ccc.length}, the length of ccc";

0.upto(dataset_size-1) do |idx|
	ccc[aaa[idx]] += 1;
end;

total = 0;
0.upto(k){ |idx| total += ccc[idx]; ccc[idx] = total;}

aaa.reverse.each { |idx| bbb[ccc[idx]] = idx; ccc[idx] -= 1;}

puts bbb;

File: books/cs/introduction-to-algorithms/chpater8/Scala/counting_sort.scala
import scala.util.Random

object CountingSort{
	def main(args: Array[String]):Unit = {
		val dataSetSize = 100
		val rnd = new Random
		val aaa = List.fill(10)(Random.nextInt(10))
		var bbb = new Array[Int](aaa.size+1)
		var ccc = new Array[Int](aaa.max+1)
		var idx = 0
		var total = 0

		println(aaa)
		println(aaa.max)
		println("the size of the array bbb")
		println(bbb.size)
		println(ccc.size)
		for( idx <- aaa ){
			println(idx)
			ccc(idx) = ccc(idx) + 1
		}
		println(ccc.toList)
		idx = 0
		total = 0
		for( idx <- 0 to (ccc.size-1) ){
			total = total + ccc(idx)
			ccc(idx) = total
		}
		println(ccc.toList)

		for(idx <- aaa.reverse){
			bbb(ccc(idx)) = idx
			ccc(idx) = ccc(idx) - 1
		}
		println(bbb.toList)
	}
}
